src/libNode/FallbackPreProcessing.cpp

```diff
@@ -0,0 +1,536 @@
+                    "be called from LookUp node.");
+        return true;
+    }
+
+    const std::multimap<NodeState, NodeState> STATE_CHECK_STATE
+        = {{WAITING_DSBLOCK, WAITING_DSBLOCK},
+           {WAITING_FINALBLOCK, WAITING_FINALBLOCK},
+           {WAITING_FALLBACKBLOCK, WAITING_FALLBACKBLOCK},
+           {MICROBLOCK_CONSENSUS, MICROBLOCK_CONSENSUS},
+           {MICROBLOCK_CONSENSUS_PREP, MICROBLOCK_CONSENSUS_PREP},
+           {MICROBLOCK_CONSENSUS_PREP, MICROBLOCK_CONSENSUS},
+           {MICROBLOCK_CONSENSUS, MICROBLOCK_CONSENSUS_PREP}};
+
+    for (auto pos = STATE_CHECK_STATE.lower_bound(nodeState);
+         pos != STATE_CHECK_STATE.upper_bound(nodeState); pos++)
+    {
+        if (pos->second == statePropose)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+// The idea of this function is to set the last know good state of the network before fallback happens.
+// This allows for the network to resume from where it left.
+void Node::SetLastKnownGoodState()
+{
+    if (LOOKUP_NODE_MODE)
+    {
+        LOG_GENERAL(WARNING,
+                    "DirectoryService::SetLastKnownGoodState not expected to "
+                    "be called from LookUp node.");
+        return;
+    }
+
+    switch (m_state)
+    {
+    case FALLBACK_CONSENSUS_PREP:
+    case FALLBACK_CONSENSUS:
+    case SYNC:
+        break;
+    default:
+        m_fallbackState = (NodeState)m_state;
+    }
+}
+
+void Node::ScheduleFallbackTimeout(bool started)
+{
+    LOG_MARKER();
+
+    cv_fallbackBlock.notify_all();
+
+    std::unique_lock<std::mutex> cv_lk(m_MutexCVFallbackBlock);
+    // if started, will use smaller interval
+    // otherwise, using big interval
+    if (!LOOKUP_NODE_MODE)
+    {
+        if (cv_fallbackBlock.wait_for(
+                cv_lk,
+                std::chrono::seconds(
+                    started ? FALLBACK_INTERVAL_STARTED
+                            : (FALLBACK_INTERVAL_WAITING * (m_myShardID + 1))))
+            == std::cv_status::timeout)
+        {
+            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
+                      "Initiated fallback" << (started ? " again" : ""));
+
+            if (m_mediator.m_ds->m_mode != DirectoryService::IDLE)
+            {
+                return;
+            }
+
+            if (started)
+            {
+                UpdateFallbackConsensusLeader();
+            }
+
+            auto func = [this]() -> void { RunConsensusOnFallback(); };
+            DetachedFunction(1, func);
+        }
+    }
+
+    if (!started)
+    {
+        if (cv_fallbackBlock.wait_for(
+                cv_lk, std::chrono::seconds(FALLBACK_INTERVAL_WAITING))
+            == std::cv_status::timeout)
+        {
+            if (m_mediator.m_ds->m_mode != DirectoryService::IDLE)
+            {
+                return;
+            }
+
+            if (m_state != FALLBACK_CONSENSUS_PREP
+                || m_state != FALLBACK_CONSENSUS)
```

Comment:
This condition checking seems always true

Reply:
Ok noted